app/src/main/java/jp/co/cyberagent/stf/Service.java

```diff
@@ -367,4 +373,69 @@ public void run() {
             }
         }
     }
+
+    /**
+     * Monitors the adb state by checking /sys/class/android_usb/android0/state
+     *
+     * If state is DISCONNECTED then IdentityActivity will be shown to allow easier identification
+     * of malfunctioning devices
+     */
+    private class AdbMonitor extends Thread {
+        // If something goes wrong you have 30 seconds to kill the STFService
+        // Using smaller numbers will basically lock the devices until usb connection is established
+        private static final int INTERVAL_MS = 5000;
+
+        @Override
+        public void run() {
+            Log.d(TAG, "Starting adb monitor thread");
+            java.lang.Process process;
+            try {
+                while(!isInterrupted()) {
+                    /**
+                     * In order for the monitor to work you need to grant DUMP permission for
+                     * the apk, e.g.
+                     *
+                     * adb -d shell pm grant jp.co.cyberagent.stf android.permission.DUMP
+                     */
+                    int dumpPermission = ContextCompat.checkSelfPermission(getApplication(), Manifest.permission.DUMP);
+                    if(dumpPermission == PackageManager.PERMISSION_GRANTED) {
+                        String[] cmd = {
+                            "/system/bin/dumpsys",
+                            "usb"
+                        };
+
+                        process = Runtime.getRuntime().exec(cmd);
+
+                        /**
+                         * If the output of the command will change then by default device will be
+                         * considered connected
+                         */
+                        String kernelStateLine = "";
```

Comment:
I would perhaps get rid of this `kernelStateLine` variable and just put `boolean disconnected = false;` here. Then change that in the loop. Seems unnecessary to capture the whole line.

Reply:
After this I send the whole line to logcat (debug). I'd prefer this to stay so that it's easier to ask people to send logs with usb state

Reply:
I've actually created another state variable instead of this. I want also to print the changes in the kernel state to logcat